contrib/epee/include/serialization/wire/wrapper/array.h

```diff
@@ -0,0 +1,161 @@
+    when reading (`max_element_count` or `min_element_size`). `C` can be `void`
+    if write-only.
+
+    This wrapper meets the requirements for an optional field; `wire::field`
+    and `wire::optional_field` determine whether an empty array must be
+    encoded on the wire. Historically, empty arrays were always omitted on
+    the wire (a defacto optional field).
+
+    The `is_array` trait can also be used, but is default treated as an optional
+    field. The trait `is_optional_on_empty` traits can be specialized to disable
+    the optional on empty behavior. See `wire/traits.h` for more ifnormation
+    on the `is_optional_on_empty` trait.
+
+    `container_type` is `T` with optional `std::reference_wrapper` removed.
+    `container_type` concept requirements:
+      * `typedef` `value_type` that specifies inner type.
+      * must have `size()` method that returns number of elements.
+    Additional concept requirements for `container_type` when reading:
+      * must have `clear()` method that removes all elements (`size() == 0`).
+      * must have either: (1) `end()` and `emplace_hint(iterator, value_type&&)`
+        or (2) `emplace_back()` and `back()`:
+        * `end()` method that returns one-past the last element.
+        * `emplace_hint(iterator, value_type&&)` method that move constructs a new
+           element.
+        * `emplace_back()` method that default initializes new element
+        * `back()` method that retrieves last element by reference.
+    Additional concept requirements for `container_type` when writing:
+      * must work with foreach loop (`std::begin` and `std::end`).
+      * must work with `boost::size` (from the `boost::range` library). */
+  template<typename T, typename C = void>
+  struct array_
+  {
+    using constraint = C;
+    using container_type = unwrap_reference_t<T>;
+    using value_type = typename container_type::value_type;
+
+    // See nested `array_` overload below
+    using inner_array = std::reference_wrapper<value_type>;
+    using inner_array_const = std::reference_wrapper<const value_type>;
+
+    T container;
+
+    constexpr const container_type& get_container() const noexcept { return container; }
+    container_type& get_container() noexcept { return container; }
+
+    //! Read directly into the non-nested array
+    container_type& get_read_object() noexcept { return get_container(); }
+
+
+    // concept requirements for optional fields
+
+    explicit operator bool() const noexcept { return !get_container().empty(); }
+    array_& emplace() noexcept { return *this; }
+
+    array_& operator*() noexcept { return *this; }
+    const array_& operator*() const noexcept { return *this; }
+
+    void reset() { get_container().clear(); }
+  };
+
+  //! Nested array case
+  template<typename T, typename C, typename D>
+  struct array_<array_<T, C>, D>
+  {
+    // compute `container_type` and `value_type` recursively
```

Comment:
So are we choosing to support nested arrays now ?

Reply:
I thought we always supported nested arrays? This occurs in the ZeroMQ code, and perhaps elsewhere. You can "get-around" this by putting a class in-between like:
```c++
struct nested { std::vector<uint32_t> inner; }
std::vector<nested> outer;
```
instead of
```c++
std::vector<std::vector<uint32_t>>
```
The complexity of the nested case is a little higher than the rest of the code, so I can understand if you think it should be removed.